scalalib/src/mill/scalalib/cosmo/CosmoModule.scala

```diff
@@ -0,0 +1,155 @@
+       |  char *all_argv[total];
+       |  memset(all_argv, 0, sizeof(all_argv));
+       |
+       |  all_argv[0] = (char*)malloc((strlen(argv[0]) + 1) * sizeof(char));
+       |  strcpy(all_argv[0], argv[0]);
+       |  ${addForkArgs}
+       |  ${addForkArgsArgv0}
+       |  all_argv[${size}] = (char*)malloc((strlen("-cp") + 1) * sizeof(char));
+       |  strcpy(all_argv[${size}], "-cp");
+       |  all_argv[${size + 1}] = (char*)malloc((strlen(argv[0]) + 1) * sizeof(char));
+       |  strcpy(all_argv[${size + 1}], argv[0]);
+       |  all_argv[${size + 2}] = (char*)malloc((strlen("${finalMainClass()}") + 1) * sizeof(char));
+       |  strcpy(all_argv[${size + 2}], "${finalMainClass()}");
+       |
+       |  int i = preargv_size;
+       |  for (int count = 1; count < argc; count++) {
+       |    all_argv[i] = (char*)malloc((strlen(argv[count]) + 1) * sizeof(char));
+       |    strcpy(all_argv[i], argv[count]);
+       |    i++;
+       |  }
+       |
+       |  all_argv[total - 1] = NULL;
+       |
+       |  const char* java_opts = getenv("JAVA_OPTS");
+       |  if (java_opts != NULL) {
+       |    const char* jdk_java_options = getenv("JDK_JAVA_OPTIONS");
+       |
+       |    if (jdk_java_options != NULL) {
+       |      const char* new_jdk_java_options = xasprintf("%s %s", jdk_java_options, java_opts);
+       |      setenv("JDK_JAVA_OPTIONS", new_jdk_java_options, 1);
+       |    } else {
+       |      setenv("JDK_JAVA_OPTIONS", java_opts, 1);
+       |    }
+       |  }
+       |
+       |  execvp("java", all_argv);
+       |  if (errno == ENOENT) {
+       |    execvp("java.exe", all_argv);
+       |  }
+       |
+       |  perror("java");
+       |  exit(EXIT_FAILURE);
+       |}
+    """.stripMargin
+  }
+
+  def cosmoCompiledLauncherScript: T[PathRef] = Task {
+    os.write(Task.dest / "launcher.c", cosmoLauncherScript())
+    os.call((
+      cosmocc().path,
+      "-mtiny",
+      "-O3",
+      "-o",
+      Task.dest / "launcher",
+      Task.dest / "launcher.c"
+    ))
+
+    PathRef(Task.dest / "launcher")
+  }
+
+  def cosmoAssembly: T[PathRef] = Task {
+    val prepend = os.read.bytes(cosmoCompiledLauncherScript().path)
+    val upstream = upstreamAssembly()
+
+    val created = Assembly.create0(
+      destJar = Task.dest / "out.jar.exe",
+      inputPaths = Seq.from(localClasspath().map(_.path)),
+      manifest = manifest(),
+      prepend = Some(prepend),
+      base = Some(upstream),
+      assemblyRules = assemblyRules
+    )
+    // See https://github.com/com-lihaoyi/mill/pull/2655#issuecomment-1672468284
+    val problematicEntryCount = 65535
+
+    if (created.entries > problematicEntryCount) {
+      Result.Failure(
+        s"""The created assembly jar contains more than ${problematicEntryCount} ZIP entries.
+           |Prepended JARs of that size are known to not work correctly.
+         """.stripMargin
+      )
+    } else {
+      Result.Success(created.pathRef)
+    }
+  }
```

Comment:
Let's re-use the existing `assembly` task; it seems only a few things need to be tweaked (the `prepend` and the `destJar` name), so those can be moved to separate tasks that `CosmoModule` overrides

Reply:
After refactoring `Assembly.create` is not used anywhere anymore, all replaced with `Assembly.create0`, but I guess we keep it around for backward compat?